bundles/org.openhab.core.persistence/src/main/java/org/openhab/core/persistence/extensions/PersistenceExtensions.java

```diff
@@ -348,10 +408,42 @@ private static void internalPersist(Item item, TimeSeries timeSeries, @Nullable
                 filter.setEndDate(ZonedDateTime.now());
             }
             filter.setOrdering(forward ? Ordering.ASCENDING : Ordering.DESCENDING);
-            filter.setPageSize(1);
-            Iterable<HistoricItem> result = qService.query(filter);
-            if (result.iterator().hasNext()) {
-                return result.iterator().next().getTimestamp();
+
+            filter.setPageSize(skipEqual ? 1000 : 1);
+            int startPage = 0;
+            filter.setPageNumber(startPage);
+
+            Iterable<HistoricItem> items = qService.query(filter);
+            Iterator<HistoricItem> itemIterator = items.iterator();
+            if (itemIterator.hasNext()) {
+                if (!skipEqual) {
+                    return itemIterator.next().getTimestamp();
+                } else {
+                    State state = item.getState();
+                    HistoricItem historicItem = itemIterator.next();
+                    int itemCount = 1;
+                    if (!historicItem.getState().equals(state)) {
+                        // Stored value different from current state, use now as timestamp when looking backward
+                        return forward ? historicItem.getTimestamp() : ZonedDateTime.now();
```

Comment:
If looking forward (nextChange) it will return the next change. If looking backward (lastChange), it will return now when the current state is different from the first one it finds in history.

Reply:
> If looking backward (lastChange), it will return now when the current state is different from the first one it finds in history.

This sounds wrong. I would've expected it to return the time when the change occurred, instead of the current time.

Reply:
> > If looking backward (lastChange), it will return now when the current state is different from the first one it finds in history.
> 
> I would've expected it to return the time when the change occurred, instead of the current time.

That’s the whole point here. If the change just occurred, it me be it was not written to persistence yet. In that case, now is the best approximation. As long as the state in persistence is the same as the current change, it will keep on looking further back in history.

Reply:
This makes me think some similar check may have to be done for last update. If current state different from last state, we know it has been updated since and may return now. That’s incomplete of course because we cannot detect if it has been updated and not stored yet.

Reply:
Thinking out loud here. Let's say item A is set to persist every change.
timestamp 1: NULL -> nothing persisted
timestamp 2: updated to `foo` -> persisted [2, "foo"]. Current state: "foo"
... time passes.....
timestamp 5: I call `lastChange("A")`. Shouldn't it here tell me that the last time item A changed state was at timestamp 2, and not timestamp 5?

Reply:
> That’s the whole point here. If the change just occurred, it may be it was not written to persistence yet. In that case, now is the best approximation. As long as the state in persistence is the same as the current change, it will keep on looking further back in history.

Ahh yes, I see this is inside the check for `if (!historicItem.getState().equals(state))` so you're correct!

Reply:
> This makes me think some similar check may have to be done for last update. If current state different from last state, we know it has been updated since and may return now. That’s incomplete of course because we cannot detect if it has been updated and not stored yet.

Added a check for state change when looking for last update.

Reply:
@mherwege I think this is a design decision. It's called "persistence extension", so I would expect that I get something from persistence. The `lastUpdate` should be the regadring the last persisted value, and `lastChange` the last persisted change. The same applies to `nextChange`. None of these should ever return the current time.

Reply:
Then it should return null in this case

Reply:
I look at this from a user perspective. If the current state is different from the last state in persistence, it is very unexpected for the result to be the time of the last value in persistence. Because obviously the state has changed since. A user in the end does not care where this information comes from. That is actually what happens with the existing methods previousState and previousUpdate. In that case, I should close this PR as this was attempting to solve that problem, and the time is part of the returned state in these other methods. Every user rule then needs to check if the returned state is different from the current state, and if so use the logic I tried to implement here.
In these persistence extensions there are various other methods where the current state is considered when returning a result. So I don’t agree with this limiting view on things.
Note that none of this applies to nextState or nextUpdate, nextChange or nextUpdate. There it will always be from persistence.

Reply:
Does this case ever actually happen, when user code would call `lastChange` after a state change occurred but before persistence extensions finished committing the change? If so, `lastChange` should either return `null` to indicate that Persistence hasn't caught up, or perhaps internally wait / yield so Persistence can catch up before returning?

Reply:
It certainly does if you do not implement an everyChange strategy. But even with an everyChange strategy, it is still not impossible. There is no guarantee on sequence of events here.

Reply:
What if the item's persistence setting is manual, or cron, and not on every change?

Reply:
Sure, and again you would expect to to see the time when the current state first appeared in the persistence (looking backwards). And if the first entry is different from the current state, it cannot be used. Returning null is an option, but it leaves the user with the problem to solve in the rule. I prefer returning the best guess.
Here is what I try to solve with PR: https://community.openhab.org/t/persistence-on-off-transitions-with-influx/156271/9?u=mherwege

Reply:
> Here is what I try to solve with PR: https://community.openhab.org/t/persistence-on-off-transitions-with-influx/156271/9?u=mherwege

This PR in general, does seem to solve the issue i.e. to give the timestamp of the first 100 (using the example data in the linked post). 



> Sure, and again you would expect to to see the time when the current state first appeared in the persistence (looking backwards). And if the first entry is different from the current state, it cannot be used, because it means the state has changed since. Returning null is an option, but it leaves the user with the problem to solve in the rule. I prefer returning the best guess.

Returning null here, tells me exactly the situation: it hasn't been persisted yet. I tend to agree with @J-N-K that the persistence extension should provide information about the persistence service and not make decisions for the user. Let the user decide what to do with the null information. Returning the current time (now) is not correct anyway. What if the change occurred 40 minutes ago, but the user only has a hourly cron persistence schedule? It's the user's fault for not having `everyChange` policy in this case but returning `now` doesn't highlight that issue.

In the case for a brief timing issue, maybe: calling yield / sleep for 100ms then retry (once) might solve that?

I wonder if we could prioritise the listeners from the persistence extensions so they get called first, and therefore persist the change first, before listeners for rule triggers are called. At the moment I believe they are non deterministic, right?

Reply:
> What if the change occurred 40 minutes ago, but the user only has a hourly cron persistence schedule? It's the user's fault for not having `everyChange` policy in this case but returning `now` doesn't highlight that issue.

A user may have decided hour resolution is good enough for his use case. Returning null would mean we refuse to answer the simple question of when the last change happend within hour resolution.

if you look at the persistence extensions in general, there are various places where we use the current state with timestamp now to return a reasonable answer. Have al look at all the minimum, maximum and deviations methods where we even construct a new historic item with current timestamp. Not doing it here to solve something which is an obvious functional lack (not being able totell since when we have the current state) looks funny to me.

Reply:
> Returning null would mean we refuse to answer the simple question of when the last change happend within hour resolution.

We should indeed refuse to answer because we (the _persistence extension_, which provides this method) do not have the requested information. The caller of this method relies on extracting the information that is stored in the persistence database, and not some made up data. Returning `now` will cause confusion because that doesn't match what's stored in the persistence database.

It should be easy enough, hopefully, for the user script to say `lastChange = PersistenceExtensions.lastChange(item) || ZonedDateTime.now()` (syntax subject to the language used) if that were the intention of the code.

Reply:
Let's make an example of what happens if we takle the current value into account.

The item has a "every five minutes" startegy configured. The last persisted value has a timestamp 10:00:00 (and did change from previous value).

10:01:10 `lastChange` returns 10:00:00
10:02:10 `lastChange` returns 10:00:00
10:02:20 value is changed
10:03:10 `lastChange` returns 10:03:00
10:04:10 `lastChange` returns 10:04:00
10:05:00 value is persisted
10:05:10 `lastChange` returns 10:05:00
10:06:10 `lastChange` returns 10:05:00

The value did not change since 10:02:20, but the persistence extension returns three different values for the same change. This looks confusing.

Reply:
I know that, and I actually prefer this inconsistency over any other solution. Bu I get the consensus is to rather return null when the current state is different from the last persisted state. If that’s the preference so be it. It is not mine, but I will adapt.
Any return value will always depend on the logic, persistence strategy and specific behavior of the persistence service. If you continue your line of thinking, the persistence extensions should also return null for any since calculation when the current state is different from the last state. You never know exactly what happened between two persisted states. Calculations could actually only be allowed on a service that does not aggregate over time (so not on rrd4j) and only when there is an every change persistence strategy. All else is wrong anyway. And as data gets aggregated over time, it gets worse. We always accepted that. So while I think it is very inconsistent not to allow this approximation and gives a worse user experience, I will accept and adapt the code and javadoc.

Reply:
@J-N-K @jimtng I adjusted the PR to return null when the current state is different and extended the description of this PR to make things more clear.